### DISCUSSION TITLE
New RxLoader

### DISCUSSION THREAD

**Discussion by gswifort** [CEXT-Dan/PyRx#146] [Ideas]
_Based on issue #145_

> Remove special VENV handling from the main module, and loader, should behave as any other python package. Users will be responsible for what module gets loaded into CAD, in a VENV context.

So we assume that the loader (e.g. RxLoader25.0.arx) and the main module (e.g. PyRx25.0.arx) will be in python packages (...\site-packages\pyrx\...) and based on the current environment (whether global or venv) the appropriate main module will be loaded. In my opinion the loader should not be tightly linked to the main module. I assume that the loader can be permanently added to the CAD start list and based on the python environment context the appropriate main module will be loaded. I think that after installing `pyrx` via `pip` there should be a command `python -m pyrx install-global` which will copy the loader to `%localappdata%\Programs\PyRx` and this loader will be permanently linked to the CAD start list.

**Comment by CEXT-Dan** [comment]
I would like to remove %localappdata%\Programs\PyRx and put the entire package into ..\site-packages\pyrx
Guessing this will break how loading from a VENV works now.  What alternatives are there? 

One option is you use CAD profiles for VENV? Does ZwCAD have Command Line Switch?
Like https://help.autodesk.com/view/ACD/2023/ENU/?guid=GUID-8E54B6EC-5B52-4F62-B7FC-0D4E1EDF093A

Maybe a “global” loader could be another project?

**Reply by gswifort** [reply]
Yes, ZWCAD has command line switches. I don't know if they all do, I tested `/b` (Script name).

I don't know if putting the loader in a separate project is a good idea, can the main module work without the loader if loaded manually?

The loader can be placed anywhere, %localappdata% is probably the most sensible place - it's not about keeping the whole project there, just the loaders, but the loader can just as well be installed in the site-packages of the global python installation (or even in venv) and the user will decide which loader to attach to the CAD start list.

I think the loader would have to check if the %VIRTUAL_ENV% variable is set

**Reply by gswifort** [reply]
After some thought, I don't know if removing %localappdata%\Programs\PyRx is a good idea.

I think there are two scenarios to consider:

1. Automation (using CAD as an engine - usually used by people who know what they are doing). In this scenario, CAD can be launched in a specific environment, with a specific master module loaded, etc. This can be done using .scr scripts (`/b` switch), lisp scripts (run from a .scr script)

2. CAD extensions - In this scenario, the extension developer may want to rely on the main module being loaded and the main module loading its bundle into CAD if it is placed in a specific directory (e.g. %localappdata%\Programs\PyRx\bundle). PyRevit can be used as an example here. Of course, in this scenario, there may be a problem with conflicting dependencies of individual extensions (which should be addressed in a separate thread).

**Reply by CEXT-Dan** [reply]
If I keep %localappdata%\Programs\PyRx, then I should keep the installer, so it can all be uninstalled, but then it makes no sense to continue with #145.  The whole idea was to simplify things by removing directory searches, not add them.

**Reply by gswifort** [reply]
I think that what I am writing about in scenario number 2 can be implemented entirely in python (I can take care of it). pyrx will remain a package installed by pip, but with an additional option to perform a global installation to %localappdata% (but not via the MSI installer but via `python -m pyrx install-global`) - thanks to this, the extension developer will be able to check if the global installation has been performed and if not, he will simply perform it from the extension installer. So the loader should probably load the main module located in the same folder. Even if the global loader is added to the start list, running CAD with the /ld switch and specifying a specific loader will cause the specified module to have priority.

**Comment by CEXT-Dan** [comment]
Ok, I’ll start by making the loader work from anywhere, I.e. ..\site-packages\pyrx,  %localappdata%\Programs\PyRx,  or in an AutoCAD bundle
I would like to have PIP -uninstall remove %localappdata%\Programs\PyRx if its created with python -m pyrx install-global

**Reply by gswifort** [reply]
Removing a global install via pip uninstall may not be possible, but I'll see if I can define additional scripts to execute during uninstall.

Are we going to publish the package on PyPI?

**Reply by gswifort** [reply]
After a short search I'm almost sure that it's not possible to remove a global installation via pip uninstall, you need to define an additional script that the user will have to execute directly.

**Reply by CEXT-Dan** [reply]
Ouch, yeah, that’s going to be a problem, I think it’s bad practice to leave residual files on a user’s machine from an install. 
Regarding PyPi, would be nice, but we don’t have a clear path using PIP yet. I did email PyRx about the name though.

I’ve had a company reach out to me with regards to support with this project, I want to have a clear path to install / uninstall, that could be rolled out company wide.  
I need to reach out to the CAD Manager’s group for advice.

**Reply by gswifort** [reply]
`pip install` is one procedure, the reverse of which is `pip uninstall`. I think it's not bad practice to undo `python -m pyrx install-global` with `python -m pyrx uninstall-global`.

**Reply by CEXT-Dan** [reply]
This will also mean there will be a need for update-global? Do you see any probability of versioning issues?

**Reply by gswifort** [reply]
Yes, but..

It doesn't have to be `update-global`, the update can be done with the same command (`install-global`).
This shouldn't cause any problems.

**Comment by CEXT-Dan** [comment]
PyRxLoader, in the global path seems to load VENV
If I have BricsCAD set to autoload in global ..\site-packages\pyrx
Activate a VENV, then run & 'C:\Program Files\Bricsys\BricsCAD V25 en_US/Bricscad.exe'
It loads the VENV copy of PyRxV25.0.brx
I think this will remove the need for having %localappdata%\Programs\PyRx and the extra supporting PIP commands

```
-----------------------------
Log:
PyRxLoader::getPythonVenvPath E:\TEST123\venv
PyRxLoader::thisModulePath M:\Dev\Projects\PyRxGit\pyrx\
PyRxLoader::getInstallPath E:\TEST123\venv\Lib\site-packages\pyrx
PyRxLoader::tryFindPythonPath c:\users\dan\appdata\local\programs\python\python312
PyRxLoader::tryFindWxPythonPath c:\users\dan\appdata\local\programs\python\python312\Lib\site-packages\wx
Loading PyRx from venv condition
PyRxLoader::PyRxLoader_loader Loading, E:\TEST123\venv\Lib\site-packages\pyrx\PyRxV25.0.brx
PyRxLoader_loader SUCCESS

-----------------------------
PyRx version <2.1.001.20241224> loaded:
Python Interpreter Loaded successfully!
: PYCMDPROMPT
>>>: import sys
>>>: print(*sys.path, sep="\n")
c:\users\dan\appdata\local\programs\python\python312\python312.zip
c:\users\dan\appdata\local\programs\python\python312\DLLs
c:\users\dan\appdata\local\programs\python\python312\Lib
C:\Users\Dan\AppData\Local\Programs\Python\Python312
E:\TEST123\venv
E:\TEST123\venv\Lib\site-packages
E:\TEST123\venv\Lib\site-packages\pyrx\
>>>: print(sys.executable)
E:\TEST123\venv\Scripts\python.exe

```

**Comment by gswifort** [comment]
![image](https://github.com/user-attachments/assets/ff5e3408-243f-4885-b6db-b61b632684ce)

here:

https://github.com/gswifort/PyRx/blob/2945f9fd5822233454e382005da9c21da1aa7573/PyRxLoader/acrxEntryPoint.cpp#L299-L318

it's missing:

https://github.com/gswifort/PyRx/blob/2945f9fd5822233454e382005da9c21da1aa7573/PyRxLoader/acrxEntryPoint.cpp#L256-L264

**Reply by CEXT-Dan** [reply]
I see, I'll see if I can fix that

**Reply by CEXT-Dan** [reply]
Try it now, I need to refactor this at some point

**Reply by gswifort** [reply]
After quick tests - it works.
Thank you!

**Reply by gswifort** [reply]
I would be grateful for the release.

**Reply by CEXT-Dan** [reply]
cad-pyrx 2.2.2
